org/postgresql/core/PGStream.java

```diff
@@ -45,37 +50,101 @@
-     * @exception IOException if an IOException occurs below it.
-     */
-    public PGStream(HostSpec hostSpec, int timeout) throws IOException
+    private final SocketFactory socketFactory;
+
+//    /**
+//     * Constructor:  Connect to the PostgreSQL back end and return
+//     * a stream connection.
+//     *
+//     * @param hostSpec the host and port to connect to
+//     * @param timeout timeout in milliseconds, or 0 if no timeout set
+//     * @exception IOException if an IOException occurs below it.
+//     */
+//    public PGStream(HostSpec hostSpec, int timeout) throws IOException
+//    {
+//        this.hostSpec = hostSpec;
+//
+//        Socket socket = new Socket();
+//        socket.connect(new InetSocketAddress(hostSpec.getHost(), hostSpec.getPort()), timeout);
+//        changeSocket(socket);
+//        setEncoding(Encoding.getJVMEncoding("US-ASCII"));
+//
+//        _int2buf = new byte[2];
+//        _int4buf = new byte[4];
+//    }
+//
+//    /**
+//     * Constructor:  Connect to the PostgreSQL back end and return
+//     * a stream connection.
+//     *
+//     * @param hostSpec the host and port to connect to
+//     * @throws IOException if an IOException occurs below it.
+//     * @deprecated use {@link #PGStream(org.postgresql.util.HostSpec, int)}
+//     */
+//    public PGStream(HostSpec hostSpec) throws IOException {
+//      this(hostSpec, 0);
+//    }
+
+    public PGStream(HostSpec hostSpec, int timeout, SocketFactory socketFactory) throws IOException
     {
         this.hostSpec = hostSpec;
 
         Socket socket = new Socket();
         socket.connect(new InetSocketAddress(hostSpec.getHost(), hostSpec.getPort()), timeout);
-        changeSocket(socket);
+        if (socketFactory == null)
+			socketFactory = SocketFactory.getDefault();
+        this.socketFactory = socketFactory;
+        changeSocket(this.socketFactory.createSocket(hostSpec.getHost(), hostSpec.getPort()));
         setEncoding(Encoding.getJVMEncoding("US-ASCII"));
 
         _int2buf = new byte[2];
         _int4buf = new byte[4];
     }
 
-    /**
-     * Constructor:  Connect to the PostgreSQL back end and return
-     * a stream connection.
-     *
-     * @param hostSpec the host and port to connect to
-     * @throws IOException if an IOException occurs below it.
-     * @deprecated use {@link #PGStream(org.postgresql.util.HostSpec, int)}
-     */
-    public PGStream(HostSpec hostSpec) throws IOException {
-      this(hostSpec, 0);
+    public PGStream(HostSpec hostSpec, int timeout, Properties info) throws IOException
+    {
+        this(hostSpec, timeout, createSocketFactory(info));
+    }
+
+    public static SocketFactory createSocketFactory(Properties info) {
+        SocketFactory factory;
+
+        // Use the default factory if no specific factory is requested
+        //
+        String classname = info.getProperty("socketfactory");
+        if (classname == null)
+			factory = SocketFactory.getDefault();
+		else
```

Comment:
It's better to drop `else` here and just `return SocketFactory.getDefault()` in the previous line